psi4/src/psi4/libfock/DirectDFJ.cc

```diff
@@ -0,0 +1,324 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Psi4: an open-source quantum chemistry software package
+ *
+ * Copyright (c) 2007-2023 The Psi4 Developers.
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "psi4/libqt/qt.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/mintshelper.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/lib3index/dftensor.h"
+#include "jk.h"
+#include "SplitJK.h"
+
+#include <unordered_set>
+#include <vector>
+#include <map>
+#include <algorithm>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+using namespace psi;
+
+namespace psi {
+
+DirectDFJ::DirectDFJ(std::shared_ptr<BasisSet> primary, std::shared_ptr<BasisSet> auxiliary, Options& options) : SplitJK(primary, options), auxiliary_(auxiliary) {
+    timer_on("DirectDFJ: Setup");
+
+    // => General Setup <= //
+
+    // thread count
+    nthreads_ = 1;
+#ifdef _OPENMP
+    nthreads_ = Process::environment.get_n_threads();
+#endif
+
+    // pre-compute coulomb fitting metric
+    timer_on("DirectDFJ: DIRECTDFJ Coulomb Metric");
+
+    FittingMetric J_metric_obj(auxiliary_, true);
+    J_metric_obj.form_fitting_metric();
+    J_metric_ = J_metric_obj.get_metric();
+
+    timer_off("DirectDFJ: DIRECTDFJ Coulomb Metric");
+
+    timer_off("DirectDFJ: Setup");
+}
+
+DirectDFJ::~DirectDFJ() {}
+
+size_t DirectDFJ::num_computed_shells() {
+    return num_computed_shells_;
+}
+
+void DirectDFJ::print_header() const {
```

Comment:
Machinery pulled from the previous CompositeJK::print_DirectDFJ_header() function.